pkg/tfbridge/provider.go

```diff
@@ -568,88 +568,83 @@ func validateProviderConfig(
-	//	Label: fmt.Sprintf("%s.news", label), KeepUnknowns: true, SkipNulls: true})
-	//if err != nil {
-	//	return nil, err
-	//}
-	//config, err := MakeTerraformConfig(nil, news, r.Schema, p.info.Config, nil, p.configValues, false)
-	//if err != nil {
-	//	return nil, errors.Wrapf(err, "preparing %s's new property state", urn)
-	//}
-	//diff, err := r.Diff(state, config, nil)
-	//if err != nil {
-	//	return nil, errors.Wrapf(err, "diffing %s", urn)
-	//}
-	//
-	//detailedDiff := makeDetailedDiff(r.Schema, p.info.Config, olds, news, diff)
-	//
-	//var replaces []string
-	//var changes pulumirpc.DiffResponse_DiffChanges
-	//var properties []string
-	//hasChanges := len(detailedDiff) > 0
-	//if hasChanges {
-	//	changes = pulumirpc.DiffResponse_DIFF_SOME
-	//	for k := range detailedDiff {
-	//		// Turn the attribute name into a top-level property name by trimming everything after the first dot.
-	//		if firstSep := strings.IndexAny(k, ".["); firstSep != -1 {
-	//			k = k[:firstSep]
-	//		}
-	//		properties = append(properties, k)
-	//		if p.info.Config != nil && p.info.Config[k] != nil {
-	//			config := p.info.Config[k]
-	//			// now is it a ForceNew or not
-	//			if config.ForceNew != nil && *config.ForceNew {
-	//				replaces = append(replaces, k)
-	//			} else {
-	//				properties = append(properties, k)
-	//			}
-	//		}
-	//	}
-	//} else {
-	//	changes = pulumirpc.DiffResponse_DIFF_NONE
-	//}
-	//
-	//return &pulumirpc.DiffResponse{
-	//	Changes:         changes,
-	//	Replaces:        replaces,
-	//	Diffs:           properties,
-	//	DetailedDiff:    detailedDiff,
-	//	HasDetailedDiff: true,
-	//}, nil
+// A DiffConfig implementation enables the Pulumi provider to detect when provider configuration is
+// changing and suggest a replacement plans when some properties require a replace. Replacing a
+// provider then replaces all resources provisioned against this provider. See also the section on
+// Explicit Provider Configuration in the docs:
+//
+// https://www.pulumi.com/docs/concepts/resources/providers/
+//
+// There is no matching context in Terraform so this remains a Pulumi-level feature.
+func (p *Provider) DiffConfig(
+	ctx context.Context, req *pulumirpc.DiffRequest,
+) (*pulumirpc.DiffResponse, error) {
+	urn := resource.URN(req.GetUrn())
+	label := fmt.Sprintf("%s.DiffConfig(%s)", p.label(), urn)
+	glog.V(9).Infof("%s executing", label)
+
+	return plugin.NewProviderServer(&configDiffer{
+		schemaMap:   p.config,
+		schemaInfos: p.info.Config,
+	}).DiffConfig(ctx, req)
+}
+
+type configDiffer struct {
+	plugin.UnimplementedProvider
+	schemaMap   shim.SchemaMap
+	schemaInfos map[string]*SchemaInfo
+}
+
+// Schema may specify that changing a property requires a replacement. This honors overrides to
+// ProviderInfo.Schema[x].ForceNew as well as raw schema.ForceNew(), although at the moment
+// realistic TF Providers do not use ForceNew for provider-level properties.
+func (p *configDiffer) isForceNew(path resource.PropertyPath) bool {
```

Comment:
We aren't using `ForceNew` here.

```suggestion
func (p *configDiffer) shouldForceReplace(path resource.PropertyPath) bool {
```